src/main/java/org/apache/sysml/hops/estim/EstimatorMatrixHistogram.java

```diff
@@ -27,182 +27,178 @@
 import org.apache.sysml.runtime.matrix.data.SparseBlock;
 
 /**
- * This estimator implements a remarkably simple yet effective
- * approach for incorporating structural properties into sparsity
- * estimation. The key idea is to maintain row and column nnz per
- * matrix, along with additional meta data.
+ * This estimator implements a remarkably simple yet effective approach for
+ * incorporating structural properties into sparsity estimation. The key idea is
+ * to maintain row and column nnz per matrix, along with additional meta data.
  */
-public class EstimatorMatrixHistogram extends SparsityEstimator
-{
-	//internal configurations
+public class EstimatorMatrixHistogram extends SparsityEstimator {
+	// internal configurations
 	private static final boolean DEFAULT_USE_EXCEPTS = true;
-	
+
 	private final boolean _useExcepts;
-	
+
 	public EstimatorMatrixHistogram() {
 		this(DEFAULT_USE_EXCEPTS);
 	}
-	
+
 	public EstimatorMatrixHistogram(boolean useExcepts) {
 		_useExcepts = useExcepts;
 	}
-	
+
 	@Override
 	public double estim(MMNode root) {
-		//recursive histogram computation of non-leaf nodes
-		if( !root.getLeft().isLeaf() )
-			estim(root.getLeft()); //obtain synopsis
-		if( !root.getRight().isLeaf() )
-			estim(root.getLeft()); //obtain synopsis
-		MatrixHistogram h1 = !root.getLeft().isLeaf() ?
-			(MatrixHistogram)root.getLeft().getSynopsis() :
-			new MatrixHistogram(root.getLeft().getData(), _useExcepts);
-		MatrixHistogram h2 = !root.getRight().isLeaf() ?
-			(MatrixHistogram)root.getRight().getSynopsis() :
-			new MatrixHistogram(root.getRight().getData(), _useExcepts);
-		
-		//estimate output sparsity based on input histograms
+		// recursive histogram computation of non-leaf nodes
+		if (!root.getLeft().isLeaf())
+			estim(root.getLeft()); // obtain synopsis
+		if (!root.getRight().isLeaf())
+			estim(root.getLeft()); // obtain synopsis
+		MatrixHistogram h1 = !root.getLeft().isLeaf() ? (MatrixHistogram) root.getLeft().getSynopsis()
+				: new MatrixHistogram(root.getLeft().getData(), _useExcepts);
+		MatrixHistogram h2 = !root.getRight().isLeaf() ? (MatrixHistogram) root.getRight().getSynopsis()
+				: new MatrixHistogram(root.getRight().getData(), _useExcepts);
+
+		// estimate output sparsity based on input histograms
 		double ret = estimIntern(h1, h2);
-		
-		//derive and memoize output histogram
+
+		// derive and memoize output histogram
 		root.setSynopsis(MatrixHistogram.deriveOutputHistogram(h1, h2, ret));
-		
+
 		return ret;
 	}
 
 	@Override
 	public double estim(MatrixBlock m1, MatrixBlock m2) {
 		MatrixHistogram h1 = new MatrixHistogram(m1, _useExcepts);
 		MatrixHistogram h2 = new MatrixHistogram(m2, _useExcepts);
-		return estimIntern(h1, h2);
+		return Math.min(estimIntern(h1, h2), new EstimatorBasicWorst().estim(m1, m2));
```

Comment:
In general, this is a good idea. However, we can do even better by exploiting the number of rows and columns with non-zero counts. Please materialize this information (nnzR, nnzC) in the histogram. In estimIntern, we can then simply do a similar min(nnz, (long)h1._nnzR*h2._nnzC) before translating it to the sparsity.